src/axom/quest/DistributedClosestPoint.hpp

```diff
@@ -0,0 +1,1093 @@
+ * \param [in] request An instance of ISendRequest that holds state for the sent data
+ * \note Adapted from conduit's relay::mpi's \a send_using_schema and \a isend to use
+ * non-blocking \a MPI_Isend instead of blocking \a MPI_Send
+ */
+inline int isend_using_schema(conduit::Node& node,
+                              int dest,
+                              int tag,
+                              MPI_Comm comm,
+                              ISendRequest* request)
+{
+  // schema will only be valid if compact and contig
+  if(node.is_compact() && node.is_contiguous())
+  {
+    request->node_schema = node.schema();
+  }
+  else
+  {
+    node.schema().compact_to(request->node_schema);
+  }
+  const std::string snd_schema_json = request->node_schema.to_json();
+
+  // create a compact schema to use
+  conduit::Schema s_msg;
+  s_msg["schema_len"].set(conduit::DataType::int64());
+  s_msg["schema"].set(conduit::DataType::char8_str(snd_schema_json.size() + 1));
+  s_msg["data"].set(request->node_schema);
+  s_msg.compact_to(request->msg_schema);
+  request->msg_node.reset();
+  request->msg_node.set_schema(request->msg_schema);
+
+  // set up the message's node using this schema
+  request->msg_node["schema_len"].set((int64)snd_schema_json.length());
+  request->msg_node["schema"].set(snd_schema_json);
+  request->msg_node["data"].update(node);
+
+  auto msg_data_size = request->msg_node.total_bytes_compact();
+  int mpi_error = MPI_Isend(const_cast<void*>(request->msg_node.data_ptr()),
+                            static_cast<int>(msg_data_size),
+                            MPI_BYTE,
+                            dest,
+                            tag,
+                            comm,
+                            &(request->mpi_request));
+
+  // Error checking -- Note: expansion of CONDUIT_CHECK_MPI_ERROR
+  if(static_cast<int>(mpi_error) != MPI_SUCCESS)
+  {
+    char check_mpi_err_str_buff[MPI_MAX_ERROR_STRING];
+    int check_mpi_err_str_len = 0;
+    MPI_Error_string(mpi_error, check_mpi_err_str_buff, &check_mpi_err_str_len);
+
+    SLIC_ERROR(
+      fmt::format("MPI call failed: error code = {} error message = {}",
+                  mpi_error,
+                  check_mpi_err_str_buff));
+  }
+
+  return mpi_error;
+}
+
+/**
+ * \brief Orchestrates sending and receiving conduit nodes among mpi ranks
+ *
+ * \param [in] send_node rooted node to send
+ * \param [in] recv_node rooted node to receive
+ * \param [in] send_rank MPI rank to send \a send_node to
+ * \param [in] recv_rank MPI rank to receive \a recv_node from
+ * \param [in] tag tag for MPI messages
+ * \param [in] comm MPI communicator to use
+ * 
+ * Sends/receives a pair of conduit nodes (along with their schemas) from the current rank.
+ * Uses non-blocking send (isend) and blocking receives and ensures that the request
+ * objects associated with the send are finalized
+ */
+inline void send_and_recv_node(conduit::Node& send_node,
```

Comment:
Have you taken a look at the `communicate_using_schema` helper class in Relay for setting up isends/irecvs?

Reply:
Thanks @kanye-quest -- that looks promising.

@gunney1, when you're improving the MPI part of this, please look into if we can use and/or adapt `communicate_using_schema` to schedule the async send/receives.

https://github.com/LLNL/conduit/blob/3edfd63095da3e7a94ea2ae4537abb5ae7b4b547/src/libs/relay/conduit_relay_mpi.hpp#L256-L266